lib/wasi/src/runtime/mod.rs

```diff
@@ -219,7 +202,7 @@ impl WasiRuntime for PluggableRuntime {
     }
 
     fn package_loader(&self) -> Arc<dyn PackageLoader + Send + Sync> {
-        Arc::clone(&self.loader)
+        Arc::clone(&self.package_loader)
```

Comment:
Don't want to create a new `Arc` each time,  should be an `Arc` on the field already.

I know this means we need to store an `Arc<dyn>`, but that should be fine?

Reply:
Besides not being able to create instances on-demand, returning a borrowed `Arc` can be annoying because you need to store *exactly* an `Arc<dyn PackageLoader + Send + Sync>`.

Something I wanted to do a couple times is store a `Arc<ConcreteType>` as a field and let the method call unsize to `Arc<dyn PackageLoader + Send + Sync>`, and that's impossible when you store a reference.

Here's a dummy piece of code that shows what I mean:

```rust
use std::sync::Arc;

struct Runtime {
    registry: Arc<MyRegistry>,
}

impl Runtime {
    fn get_registry(&self) -> Arc<dyn Registry + Send + Sync> {
        self.registry.clone()
    }
}

trait Registry {}

struct MyRegistry {}

impl Registry for MyRegistry {}

impl MyRegistry {
    fn access_some_custom_method(&self) {}
}

fn main() {
    let runtime = Runtime {
        registry: Arc::new(MyRegistry {}),
    };

    runtime.registry.access_some_custom_method();

    let registry = runtime.get_registry();
}
```

[(playground)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e71b005621ed2f81a311d7b83216adab)

Cloning an `Arc` is essentially free compared to everything else we are doing. It uses a [`Ordering::Relaxed`](https://github.com/rust-lang/rust/blob/c2ccc855e74aec03e434405eca3c247ee2432e53/library/alloc/src/sync.rs#LL1501C66-L1501C66) for the increment, loading packages is multiple orders of magnitude more expensive even in the happy path.